Durable workflows

skozina opened the pull request:
This depends, or rather contains, also changes included in #17589 and #17462.

skozina committed:
api: Add durable-workflows API extension

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
shared/api: Introduce durable operation class

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/db/cluster: Add directive to get operations by NodeId and Class

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Extract operationWaitHandler() run hook into separate function

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/cluster/heartbeat: Add state argument to Send()

Durable operations will need to reconstruct the operations when heartbeats are
lost. The state.State object will be needed for the reconstruction.

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd: Introduce durable operation hooks map

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/db/cluster/operations: Introduce `UpdateOperationNodeID()`

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Introduce functions to restart durable operation

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Enhance wait handler

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina commented on lxd/operations.go:
```diff
@@ -387,6 +387,110 @@ func operationCancel(ctx context.Context, s *state.State, projectName string, op
+		op.Class = (operations.OperationClass)(dbOp.Class).String()
+		op.Description = dbOp.Description
+		op.CreatedAt = dbOp.CreatedAt
+		op.Status = api.Frozen.String() // As we are loading the operation from the DB, it's likely not running right now.
```
I don't think so. We know that this operation is not currently running on the node, or at least the node is not responsive. We could update the operation status in the DB first (once the operation is restarted, we'd set it again to Running), but that would mean we'd be writing to the database as part of the `operationsGet()` which would be confusing. I'd leave this as it is.

skozina commented on lxd/operations/linux.go:
```diff
@@ -62,3 +143,90 @@ func (op *Operation) sendEvent(eventMessage any) {
+
+			if dbOp.IdentityID != nil {
+				identityFilter := cluster.IdentityFilter{ID: dbOp.IdentityID}
+				clusterIdentities, err := cluster.GetIdentitys(ctx, tx.Tx(), identityFilter)
```
This is generated function, not worth fixing the generated for proper English naming.

skozina commented on lxd/operations/operations.go:
```diff
@@ -310,19 +329,28 @@ func (op *Operation) Start() error {
+	if err != nil {
+		op.lock.Unlock()
+		return fmt.Errorf("failed to update Operation %s status: %w", op.id, err)
+	}
```
This doesn't seem right. The last sentence seems plain wrong:
> the function returns at line 335 without ever releasing the lock acquired at line 326.

As the line before 335 looks like:
```
332		err := op.updateStatus(api.Running)
333		if err != nil {
334			op.lock.Unlock()
335			return fmt.Errorf("failed to update Operation %s status: %w", op.id, err)
336		}
```

kadinsayani commented on lxd/db/operationtype/operation_type.go:
```diff
@@ -211,6 +212,8 @@ func (t Type) Description() string {
 	case RemoveExpiredOIDCSessions:
 		return "Remove expired OIDC sessions"
+	case Wait:
+		return "Just chilling"
```
Nice 😎

kadinsayani commented on lxd/operations/operations.go:
```diff
@@ -310,19 +329,28 @@ func (op *Operation) Start() error {
 				op.err = err
+				err = op.updateStatus(api.Failure)
+				if err != nil {
+					op.logger.Warn("Failed to update operation status to failure", logger.Ctx{"err": err})
```
Can we include more info in the context of these log messages?


```suggestion
					op.logger.Warn("Failed to update operation status", logger.Ctx{"err": err, "oldStatus": op.status.String(), "newStatus": api.Failure.String()})
```

kadinsayani commented on lxd/operations/operations.go:
```diff
@@ -333,7 +361,11 @@ func (op *Operation) Start() error {
-			op.status = api.Success
+			err = op.updateStatus(api.Success)
+			if err != nil {
+				op.logger.Warn("Failed to update operation status to success", logger.Ctx{"err": err})
```
Same note on adding more context to these log messages. Since we update the status and construct these warning messages multiple times, perhaps we can extract a helper function:



```Go
func (op *Operation) updateStatusWithWarning(status api.Status) {
	err := op.updateStatus(status)
    if err != nil {
		op.logger.Warn("Failed to update operation status", logger.Ctx{
			"err":    err,
            "oldStatus": op.status.String(),
			"newStatus": status.String(),
		})
	}
}
```

Or, we can expand `updateStatus()` to handle any `error` returned by `updateDBOperationStatus()` and log warnings.

kadinsayani commented on lxd/operations/operations.go:
```diff
@@ -719,3 +789,26 @@ func (op *Operation) Class() OperationClass {
+// Durable operations need to have handlers defined in a durableOperations map rather than provided via arguments.
+// Arguments to these handlers are then looked up in the metadata.
+func CreateDurableOperation(ctx context.Context, s *state.State, opUUID string, projectName string, opType operationtype.Type, opResources map[string][]api.URL, opMetadata map[string]string) (*Operation, error) {
+	handlers, ok := durableOperations[opType]
```
We need a synchronization mechanism to prevent concurrent access here too.

kadinsayani replied:
Ack, thanks for clarifying.

kadinsayani reviewed (changes requested):
Looking really good so far! I've made an initial pass through and have a few questions, comments and suggestions.

kadinsayani commented on lxd/operations/operations.go:
```diff
@@ -719,3 +789,26 @@ func (op *Operation) Class() OperationClass {
+// Because this table needs to be accessible from the operations package, we provide this Init function to set it.
+func InitDurableOperations(opTable map[operationtype.Type]DurableOperationHandlers) {
+	durableOperations = opTable
+}
```
Ah, I just looked at the call site of `InitDurableOperations()`. Since `d.init()` is thread safe, we're all good! Keeping this unresolved as the type suggestion is still worth considering (`type DurableOperationTable map[operationtype.Type]DurableOperationHandlers`).

kadinsayani replied:
I'm not sure if the use of `sync.Once()` is strictly required here since `daemon.init()` is already thread safe, however using `sync.Once()` does make the code self-documenting by indicating it should only be created once. Will `InitDurableOperations()` be called anywhere else?

skozina commented on lxd/db/cluster/projects.go:
```diff
@@ -156,6 +156,39 @@ func GetProjectNames(ctx context.Context, tx *sql.Tx) ([]string, error) {
+	err = rows.Err()
+	if err != nil {
+		return nil, err
+	}
```
Thanks! Switched to `query.Scan()`.

skozina commented on shared/api/operation.go:
```diff
@@ -15,6 +15,9 @@ const OperationClassWebsocket = "websocket"
 // OperationClassToken represents the Token OperationClass.
 const OperationClassToken = "token"
 
+// OperationClassDurable represents the Durable OperationClass.
```
Good idea! I've added a sentence to each. Appreciate if you could check for correctness;-)

skozina commented on lxd/operations.go:
```diff
@@ -1353,10 +1520,22 @@ func operationWaitHandler(d *Daemon, r *http.Request) response.Response {
 	return operations.OperationResponse(op)
 }
+
+var DurableOperations = map[operationtype.Type]operations.DurableOperationHandlers{
```
Yup, done (just in a later commit when we're populating the table). Thanks!

markylaing commented on lxd/operations/linux.go:
```diff
@@ -43,6 +79,51 @@ func registerDBOperation(op *Operation, opType operationtype.Type) error {
 	return nil
 }
 
+func updateDBOperationNodeID(op *Operation) error {
```
I'm not a huge fan of how we're using build tags to define how operations interact with the database.

I'd like to see something like an interface defined in the operations package. When creating an operation, we can pass an implementation of that interface which does that actual persistence to the DB.

tomponline replied:
there should be no need for build tags to be needed if we can untangle server-side operations from lxd-agent operations.

skozina replied:
I see, thank you!
For now I've opened #17268 to track this. Thanks!

skozina committed:
lxd/cluster/heartbeat: Restart durable operations on failed heartbeat

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Configure hooks for operationWaitHandler()

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Add checks for durable operations to initOperation()

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/db/cluster/operations: Remove unused functions

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

kadinsayani commented on lxd/operations/linux.go:
```diff
@@ -78,3 +83,43 @@ func (op *Operation) sendEvent(eventMessage any) {
+	}
+
+	op.resources = nil
+	op.metadata = nil
```
This can be cleaned up as follows:

```Go
op := Operation{
    projectName: projectName,
    ...
}

kadinsayani commented on lxd/operations/linux.go:
```diff
@@ -55,10 +166,225 @@ func removeDBOperation(op *Operation) error {
+		filter := cluster.OperationFilter{NodeID: &nodeID, Class: &durableClass}
+		dbOps, err := cluster.GetOperations(ctx, tx.Tx(), filter)
+		if err != nil {
+			return fmt.Errorf("Failed loading durable operations for the node %d: %w", nodeID, err)
```
```suggestion
			return fmt.Errorf("Failed loading durable operations for node with ID %q: %w", nodeID, err)
```

kadinsayani commented on lxd/operations/notlinux.go:
```diff
@@ -17,9 +17,41 @@ func registerDBOperation(op *Operation, opType operationtype.Type) error {
+
+func removeDBOperationMetadata(op *Operation) error {
+	if op.state != nil {
+		return fmt.Errorf("removeOperationMetadata not supported on this platform")
```
```suggestion
		return fmt.Errorf("removeDBOperationMetadata not supported on this platform")
```

kadinsayani commented on lxd/operations/operations.go:
```diff
@@ -278,6 +342,27 @@ func (op *Operation) done() {
 	op.lock.Unlock()
 
+	// For durable operations, we just remove it from the internal map and clear metadata from the db.
+	// The database record will be cleared later by a dedicated task.
```
```suggestion
	// The database record will be cleared later by a dedicated task.
	// XXX: Add reference to task and include details on interval once the task exists.
```

kadinsayani commented on lxd/operations.go:
```diff
@@ -158,50 +165,97 @@ func operationGet(d *Daemon, r *http.Request) response.Response {
+			}
+
+			if len(ops) > 1 {
+				return errors.New("More than one operation matches")
```
When is this possible?

kadinsayani replied:
Perhaps a comment explaining? :)

kadinsayani commented on lxd/operations.go:
```diff
@@ -158,50 +165,97 @@ func operationGet(d *Daemon, r *http.Request) response.Response {
 
-		if len(ops) > 1 {
-			return errors.New("More than one operation matches")
+		// The operations.Operation doesn't contain the node where the operation is running.
```
```suggestion
		// The [operations.Operation] doesn't contain the node where the operation is running.
```

godoc links are handy :)

kadinsayani commented on lxd/operations.go:
```diff
@@ -314,16 +368,23 @@ func operationDelete(d *Daemon, r *http.Request) response.Response {
+			return response.BadRequest(errors.New("Durable operation is already finalized"))
+		}
+
+		return response.SmartError(fmt.Errorf("Operation %q is not running on this node", id))
```
```suggestion
		return response.SmartError(fmt.Errorf("Operation with ID%q is not running on this node", id))
```

kadinsayani commented on lxd/operations.go:
```diff
@@ -527,6 +599,83 @@ func operationsGet(d *Daemon, r *http.Request) response.Response {
+				// Durable operation, convert to API representation now.
+				op, err := operations.NewDurableOperation(ctx, tx.Tx(), s, &dbOp, projectName)
+				if err != nil {
+					return fmt.Errorf("Failed to load durable operation %q: %w", dbOp.Reference, err)
```
```suggestion
					return fmt.Errorf("Failed loading durable operation with reference %q: %w", dbOp.Reference, err)
```

skozina replied:
Thanks, fixed! The `ID` is the API external field which we internally call `Reference` (because ID is the database ID). So in this case I'm going to say `operation ID`, in line with other messages.

kadinsayani reviewed (commented):
Just a few nits, suggestions, and a couple questions :)

This is really coming together!

skozina commented on lxd/operations.go:
```diff
@@ -641,64 +798,56 @@ func operationsGet(d *Daemon, r *http.Request) response.Response {
-		var err error
+	// Merge with the durable operations data.
+	for _, o := range durableOps {
+		op := o // Local var for pointer.
```
I've copied this from existing code;-) But I think this explains it nicely:
https://stackoverflow.com/questions/38692998/strange-golang-append-behavior-overwriting-values-in-slice
```
The for ... range makes a copy of the elements it loops over, and you append the address of this temporary, loop variable - which is the same in all iterations. 
```

Without the copy of the loop variable, we'd (in the case of `recursion`) return slice of pointers all pointing to the same variable.

skozina replied:
Thank you for those references! And thanks also for cleaning up the existing code;-) I've dropped the unnecessary local variable. Thanks!

skozina committed:
lxd: Cancel durable operations on node when heartbeats are not received

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd: Introduce local task to keep local durable operations in sync with the DB

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Reload durable operation from DB when created

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Add child operation to operationWaitHandler()

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

markylaing commented on lxd/operations/linux.go:
```diff
@@ -9,15 +9,28 @@ import (
 )
 
-func registerDBOperation(op *Operation) error {
+func registerDBOperation(op *Operation, conflictReference string) error {
```
Do you need to change the signature here? The conflict reference is a field on Operation isn't it?

markylaing commented on lxd/operations.go:
```diff
@@ -1488,6 +1488,74 @@ func pruneExpiredDurableOperationsTask(stateFunc func() *state.State) (task.Func
+
+		// Ensure that all durable operations in the database are running locally.
+		for _, op := range dbOps {
+			if op.Status().IsFinal() {
```
Can we please always use `op.IsRunning`? This checks the run context directly and cannot race with subsequent state updates.

skozina replied:
I've changed the PR to use following pattern:
 - code using only `cluster.Operation`, the `operations.Operation` is not even constructed: There's no context, so we are using `status.IsFinal()` to check state
 - code where `cluster.Operation` is converted into `operations.Operation`, but not really used yet, there's no operation running: The conversion code cancels the `running` context if `status.IsFinal()`, so we are using `operation.IsRunning()` to check state, which looks at `running` context
 - code where we are looking at running operations: We are always working with `operations.Operation`, so we are always using `operation.IsRunning()` to check its state.

skozina commented on lxd/db/cluster/operations.go:
```diff
@@ -65,9 +66,10 @@ type Operation struct {
+	ID                *int64
+	NodeID            *int64
+	UUID              *string
+	ConflictReference *string
```
The durable-workflows branch contains all commits in order in which these will land in several independent PRs, at least:
 - schema changes (merged)
 - conflict reference (merged)
 - operation(s)Get returning operations constructed from DB
 - bulk operations
 - durable operations
The ConflictReference is already merged and thus is not in this PR any more.

skozina commented on lxd/db/operationtype/operation_type.go:
```diff
@@ -335,3 +335,18 @@ func (t Type) EntityType() entity.Type {
+
+// ConflictAction returns the action to take if a conflicting operation is already running.
+func (t Type) ConflictAction() ConflictAction {
+	return ConflictActionNone
```
Yes. We discussed this in the conflict reference PR:
https://github.com/canonical/lxd/pull/17721#discussion_r2840227020

skozina commented on doc/api-extensions.md:
```diff
@@ -2968,3 +2968,8 @@ As with the {ref}`storage and profile operation extension <extension-storage-and
+(extension-durable-operations)=
+## `durable_operations`
+
+Introduces new operation class for durable operations. Durable operations are guaranteed to run on a different node if the node running the operation previously goes offline.
```
This will be the first commit of the durable operations PR. All commits before this one will land in separate PRs.
Note that there's another API extension for adding the `Parent` and `Children` fields in the `api.Operation` struct. I've moved that one to also be the first commit of the bulk operations PR.

skozina commented on lxd/operations.go:
```diff
@@ -512,6 +523,81 @@ func operationsGet(d *Daemon, r *http.Request) response.Response {
+	var durableOps []api.Operation
+	membersWithOps := make(map[string]struct{})
+	var members []db.NodeInfo
+	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
```
Yes, loading all operations from the DB on `GET /1.0/operations` will be a separate PR, roughly the first 10 commits of the durable-workflows branch.
The operations are loaded on line 629 below.

skozina committed:
lxd/operations: Prune durable operations after 24 hours

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Ensure matching node_id when updating operations

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

skozina committed:
lxd/operations: Cancel local durable operation if we fail to update its status

The DB SQL will update the operation entry only on the matching node_id. If the
operation was migrated to a different node, the status update fails and we
should cancel the operation locally.

Signed-off-by: Stanislav Kozina <stanislav.kozina@canonical.com>

simondeziel commented on test/suites/clustering.sh:
```diff
@@ -7027,4 +7027,236 @@ test_clustering_acme() {
+
+  # Wait for node2 to be detected as back online.
+  for _ in $(seq 20); do
+    if [ "$(LXD_DIR="${LXD_TWO_DIR}" lxc query /1.0/cluster/members/node2 | jq --exit-status -r '.status')" = 'Online' ]; then
```
```suggestion
    if LXD_DIR="${LXD_TWO_DIR}" lxc query /1.0/cluster/members/node2 | jq --exit-status '.status == "Online"'; then
```

simondeziel commented on test/suites/clustering.sh:
```diff
@@ -7027,4 +7027,236 @@ test_clustering_acme() {
+
+  if [ "${succeeded}" = 0 ]; then
+    echo "Durable operation did not move to the leader after network partition"
+    # Clean up partition before failing
+    ip netns exec "${ns1}" iptables -F || true
```
Could those iptables rules be left in place for an easier `LXD_INSPECT=1` behavior?

simondeziel commented on test/suites/clustering.sh:
```diff
@@ -7027,4 +7027,236 @@ test_clustering_acme() {
+
+  # Wait for partitioned member to be detected as back online.
+  for _ in $(seq 20); do
+    if [ "$(LXD_DIR="${LXD_TWO_DIR}" lxc query /1.0/cluster/members/node1 | jq --exit-status -r '.status')" = 'Online' ]; then
```
```suggestion
    if LXD_DIR="${LXD_TWO_DIR}" lxc query /1.0/cluster/members/node1 | jq --exit-status '.status == "Online"'; then
```

simondeziel commented on test/suites/clustering.sh:
```diff
@@ -7027,4 +7027,236 @@ test_clustering_acme() {
+
+  # Wait for partitioned member to be detected as back online.
+  for _ in $(seq 20); do
+    if [ "$(LXD_DIR="${LXD_ONE_DIR}" lxc query "/1.0/cluster/members/${leader_name}" | jq --exit-status -r '.status')" = 'Online' ]; then
```
```suggestion
    if LXD_DIR="${LXD_ONE_DIR}" lxc query "/1.0/cluster/members/${leader_name}" | jq --exit-status '.status == "Online"'; then
```

simondeziel commented on test/suites/clustering.sh:
```diff
@@ -7027,4 +7027,236 @@ test_clustering_acme() {
+
+  if [ "${succeeded}" = 0 ]; then
+    echo "Durable operation did not move to the leader after network partition"
+    # Clean up partition before failing
+    ip netns exec "${leader_ns}" iptables -F || true
```
Same here, I wouldn't do the cleanup on failure.

simondeziel commented on test/suites/clustering.sh:
```diff
@@ -7027,4 +7027,236 @@ test_clustering_acme() {
+  kill_go_proc "$(< "${LXD_ONE_DIR}/lxd.pid")"
+  echo "Stopped leader (member 1)"
+
+  # Wait for the operation to succeed. In this time leader election should occur and the operation should restart on the new leader.
+  succeeded=0
+  for i in $(seq 60); do
+    leader="$(lxd_leader_name "${LXD_TWO_DIR}")"
+    if LXD_DIR="${LXD_TWO_DIR}" lxc query "/1.0/operations/${operation_id}" | jq --exit-status '.status == "Success" and .location == "'"${leader}"'"'; then
+      succeeded=1
+      break
+    fi
+
+    sleep 1
+  done
+
+  if [ "${succeeded}" = 0 ]; then
+    echo "Durable operation was not restarted on the newly elected leader"
+    false
+  fi
```
Could this be moved to a separated helper function?

markylaing commented on shared/api/operation.go:
```diff
@@ -6,14 +6,23 @@ import (
-const OperationClassToken = "token"
+	// OperationClassDurable represents the Durable OperationClass.
+	// These are like task operations, but are restarted on the leader if the member that was initially running the operation goes offline.
+	OperationClassDurable = "durable"
```
Break into two commits:
 - restructure and and comments to existing
 - Add types for durable class

markylaing commented on lxd/operations/db.go:
```diff
@@ -267,6 +267,15 @@ func ConstructOperationFromDB(ctx context.Context, tx *sql.Tx, s *state.State, d
 		}
 	}
 
+	if op.class == operationtype.OperationClassDurable {
```
Either:
- Durable parent with children and no run hook
- Durable operation with run hook and no children

markylaing commented on lxd/operations/operations.go:
```diff
@@ -166,6 +169,25 @@ func ScheduleServerOperation(s *state.State, args OperationArgs) (*Operation, er
 }
 
+// DurableOperationTable represents the table of durable operation hooks.
+type DurableOperationTable map[operationtype.Type]RunHook
```
move this table to the operations package. Add function to add to it.

e.g. 
```go

func init() {
	operations.RegisterDurableOperationRunHook(operationtype.ReplicatorFinalize, replicatorFinalizeDurableHook)
}
```

markylaing commented on lxd/operations/operations.go:
```diff
@@ -295,6 +301,16 @@ func scheduleOperation(s *state.State, args OperationArgs) (*Operation, error) {
 		}
 
+		if op.class == operationtype.OperationClassDurable {
+			// Durable operations must have their hooks defined in the durable operations table.
```
Unless they are a parent

markylaing commented on lxd/operations/operations.go:
```diff
@@ -361,6 +361,35 @@ func scheduleOperation(s *state.State, args OperationArgs) (*Operation, error) {
+		}
+
+		op = newOp
+	}
```
Factor this out

markylaing commented on lxd/operations/db.go:
```diff
@@ -323,3 +323,28 @@ func PruneExpiredOperations(ctx context.Context, s *state.State) error {
+// loadDurableOperationFromDB reloads a durable operation from the database based on its UUID.
+// This is only used to ease debugging to ensure that the reloaded operation is identical to the one originally created.
+func loadDurableOperationFromDB(op *Operation) (*Operation, error) {
+	var result *Operation
```
This is called in a loop :(

Fix

markylaing commented on lxd/operations/db.go:
```diff
@@ -348,3 +348,115 @@ func loadDurableOperationFromDB(op *Operation) (*Operation, error) {
 }
+
+// LoadDurableOperationsFromNode returns all durable operations from the db that exist on given node.
+func LoadDurableOperationsFromNode(ctx context.Context, s *state.State, nodeID int64) ([]*Operation, error) {
```
Separate into load and update. Then have a "load and ensure"

Set the node ID on all operations in one query.

markylaing commented on lxd/operations/db.go:
```diff
@@ -348,3 +348,115 @@ func loadDurableOperationFromDB(op *Operation) (*Operation, error) {
+// This is used to restart operations that were running on a node which failed to respond to heartbeats on other node.
+func RestartDurableOperation(s *state.State, op *Operation) {
+	// Don't restart operations which are already in a final state.
+	if !op.IsRunning() {
```
This should be looking at the operation status rather than the operation context.

tomponline commented on lxd/operations/operations.go:
```diff
@@ -711,6 +740,50 @@ func (op *Operation) cancelImmediate() {
+// CancelLocalDurableOperations stops all durable operations running on this node.
+// These operations are only removed from the local operations map, the database records are left intact.
+// The cluster leader will later restart these operations.
+func CancelLocalDurableOperations() {
```
should these wait for the cancelled operations to finish?

tomponline commented on lxd/operations/db.go:
```diff
@@ -701,3 +701,125 @@ func loadAndConstructOperationFromDB(ctx context.Context, s *state.State, opID s
+
+	operationsLock.Unlock()
+
+	op.logger.Debug("Restarting operation", logger.Ctx{"id": op.id})
```
I think this should be Info level

tomponline commented on lxd/cluster/heartbeat.go:
```diff
@@ -197,6 +198,15 @@ func (hbState *APIHeartbeat) Send(ctx context.Context, s *state.State, networkCe
 					logger.Warn("Failed creating warning", logger.Ctx{"err": err})
 				}
+
+				offlineTime := time.Now().Add(-offlineThreshold)
```
lets have a comment here saying what its doing.

tomponline replied:
otherwise its just another block of code without a label of what its for.

tomponline replied:
not a blocker - but hb.Send is a weird place to be restarting operations (and clearing warnings for that matter) in my view.

tomponline commented on lxd/operations/args.go:
```diff
@@ -202,3 +211,19 @@ func (a OperationArgs) validate(isChild bool) error {
+		a.inputs = map[InputKey]json.RawMessage{}
+	}
+
+	b, err := json.Marshal(value)
```
@markylaing just want to check here, this wont end up storing nested json inside when a.inputs map is encoded to store in the DB will it?

markylaing replied:
I'm not sure what you mean. It's a JSON object where each key is a string but the value can be anything (as if it were a `map[string]any`). The change to this data type is so that we can have a generic function which gets the key and unmarshals it back into the same type that it was set as, rather than getting back an `any` that we need to perform type assertion against.

markylaing replied:
This change also means that values are unmarshalled lazily, so the values aren't unmarshalled until they are needed.

tomponline replied:
what i mean is that we are marshalling it into `b`, and then that later gets marshalled again as part of `input` map, but i want to check this wont result in a double marshalled json string?

tomponline replied:
but what we want to end up in the DB is the equivalent of if we just marshalled a map[string]any once. Is that is what happening with RawMessage?

markylaing replied:
Yeah that's exactly what is happening :)

tomponline commented on lxd/operations.go:
```diff
@@ -1143,6 +1143,33 @@ type operationWaitPost struct {
+
+const operationInputKeyWaitHandlerDuration operations.InputKey = "duration"
+
+func waitHandlerOperationRunHook(ctx context.Context, op *operations.Operation) error {
```
missing doc block

tomponline commented on lxd/operations.go:
```diff
@@ -1143,6 +1143,33 @@ type operationWaitPost struct {
+}
+
 // operationWaitHandler creates a dummy operation that waits for a specified duration.
 func operationWaitHandler(d *Daemon, r *http.Request) response.Response {
```
can we add "testing" to this function name so its clear its for testing?

tomponline commented on lxd/operations.go:
```diff
@@ -1160,13 +1160,54 @@ func waitHandlerOperationRunHook(ctx context.Context, op *operations.Operation)
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-time.Tick(duration):
+	logger.Warnf("Starting wait handler operation for %s", duration.String())
```
we dont use Warnf any more

simondeziel replied:
Should we have a linter script to enforce this?

tomponline commented:
a few things to confirm or to fix in a follow up PR, thanks!

tomponline commented:
@markylaing 

> Durable operation did not move to the leader after network partition

markylaing commented:
> @markylaing btrfs cluster check failed:
> 
> > Durable operation did not move to the leader after network partition

:facepalm: I have never seen this in CI before and it is not happening locally.

markylaing commented:
> > @markylaing btrfs cluster check failed:
> > > Durable operation did not move to the leader after network partition
> 
> 🤦 I have never seen this in CI before and it is not happening locally.

It worked on retry. Current theory is a race with operation synchronization logic that has changed since this PR last had test runs. When an operation is internally cancelled, it's status is set to cancelled locally but not in the database. If the operation synchronization task runs, it can update the status to cancelled in the background. Since the operation is then finalized, it is never restarted.

I think I need to leave the status locally as running (because technically it is still running, just being moved to another member)

elijahgreenstein reviewed (dismissed):
Docs update LGTM

markylaing commented:
I think this should be good to go now.

tomponline reviewed (commented):
Thanks!
